### PR TITLE
CI: add slash dispatch workflow to trigger pre-commit checks with comment

### PR DESCRIPTION
**Pull request by MarcoGorelli**
This would be an on-demand bot to run pre-commit checks on a PR, which can be triggered by commenting

```
@github-actions pre-commit
```

on a pull request (see [here](https://github.com/nbQA-dev/nbQA/pull/518) for a demo).

Use case: if a PR is submitted and is good-to-go except for some linting error, we can just comment `/pre-commit-run` and have the bot fixup the errors.

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
Cool, thanks! 
That actually reminded me that we use something similar in the arrow repo, using comments to trigger actions: https://github.com/apache/arrow/blob/master/.github/workflows/comment_bot.yml. This seems to check the content of the comment, instead of using one workflow (with slash-command-dispatch) to trigger another, which seems a bit simpler to me?

**Comment by MarcoGorelli** [comment]
Thanks @jorisvandenbossche ! I didn't know you could do that, that indeed does look simpler, I'll see if I can update this to do it that way

**Comment by MarcoGorelli** [comment]
@jorisvandenbossche OK, done

If the comment-bot is now largely taken from `arrow`, my understanding is that we need to include their license at the top of the file - is this right?

**Comment by jorisvandenbossche** [comment]
> If the comment-bot is now largely taken from arrow, my understanding is that we need to include their license at the top of the file - is this right?

I don't think that that is necesaarily needed in this case. I mean, I contribute to several projects, and I am constantly applying what I tried / learned how to set up github actions in one project to other projects. Also, if you include the license, you would need to clearly state what has changed. And in the end, the main thing you copied is the line `if: startsWith(github.event.comment.body, '@github-actions autotune')` and the idea to use the pr-push action.

File: .github/workflows/comment_bot.yml [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        uses: actions/cache@v2
+        with:
+          path: |
+            ~/.cache/pre-commit
+            ~/.cache/pip
+          key: pre-commit-dispatched-${{ runner.os }}-build
+      - uses: actions/setup-python@v2
+        with:
+          python-version: 3.8
+      - name: install-pre-commit
```
Comment (jorisvandenbossche) [inline .github/workflows/comment_bot.yml:46]:
```suggestion
      - name: Install pre-commit
```

(to be consistent with other names)

File: .github/workflows/comment_bot.yml [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+on:
+  issue_comment:
+    types:
+      - created
+      - edited
+
+jobs:
+  autotune:
+    name: "Fixup pre-commit formatting"
+    if: startsWith(github.event.comment.body, '@github-actions autotune')
```
Comment (jorisvandenbossche) [inline .github/workflows/comment_bot.yml:29]: I would maybe keep a "pre-commit" specific command 

(we can always later have other bot actions, and name something combined "autotune", but for now it's just pre-commit)

File: .github/workflows/comment_bot.yml [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        with:
+          python-version: 3.8
+      - name: install-pre-commit
+        run: python -m pip install --upgrade pre-commit
+      - name: Run pre-commit
+        run: pre-commit run --all-files || (exit 0)
+      - name: Commit results
+        run: |
+          git config user.name "$(git log -1 --pretty=format:%an)"
+          git config user.email "$(git log -1 --pretty=format:%ae)"
```
Comment (jorisvandenbossche) [inline .github/workflows/comment_bot.yml:53]: Does this use the author of the last commit on the PR as commiter here? 

Maybe your original "github actions" user is more transparent in who commited it?

          git config --local user.email "41898282+github-actions[bot]@users.noreply.github.com"
          git config --local user.name "github-actions[bot]"
          git commit -m "Run pre-commit" -a

File: .github/workflows/comment_bot.yml [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+          git commit -a -m 'Autoformat/render all the things [automated commit]' || echo "No changes to commit"
```
Comment (jorisvandenbossche) [inline .github/workflows/comment_bot.yml:54]: I would also keep your original message here, something more specific to pre-commit

**Review by jorisvandenbossche — COMMENTED** [review]
Thanks for the update!

File: .github/workflows/comment_bot.yml (hunk 3 above)
Comment (jorisvandenbossche) [reply .github/workflows/comment_bot.yml:53]: Actually, using the PR author / last commiter as commit author here is maybe good. That avoids that github wants to add the bot to the "Co-authored-by", which is not needed. 
A "[automated commit]" in the commit message should be clear enough about the origin of the change.

**Comment by jorisvandenbossche** [comment]
What are your thoughts on https://github.com/pandas-dev/pandas/pull/38444#issuecomment-746988296 ?

**Comment by MarcoGorelli** [comment]
Hey Joris - sorry, I missed that comment, somehow my brain only registered the review comments - I've removed the license header as I think your assessment makes perfect sense (as someone with far less experience, I was trying to tread carefully)

**Comment by jreback** [comment]
this looks fine to me. let's give it a try.

**Comment by jreback** [comment]
@MarcoGorelli can you add somewhere the various actions that are availabel (take is in the user docs, but this one plus the backport ones could go somewhere)

**Comment by jorisvandenbossche** [comment]
Thanks!

**Pull request merged**